atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/WriteInfoPartitioner.java

```diff
@@ -54,39 +57,59 @@ public WriteInfoPartitioner(KeyValueService kvs, Supplier<Integer> numShards) {
+                            .computeIfAbsent(partition, _k -> new ArrayList<>())
+                            .addAll(writeInfo));
+        }
+        return writesPartitionedByTimestamp;
     }
 
     private Map<PartitionInfo, List<WriteInfo>> filterAndPartitionForSingleTimestamp(List<WriteInfo> writes) {
-        if (writes.stream().allMatch(writeInfo -> getStrategy(writeInfo) == SweeperStrategy.NON_SWEEPABLE)) {
-            long startTs = writes.stream().findFirst().get().timestamp();
-            return SweepQueueUtils.partitioningForNonSweepable(startTs);
+        if (writes.isEmpty()) {
+            return Map.of();
         }
-        return partitionWritesByShardStrategyTimestamp(writes);
-    }
 
-    private Map<PartitionInfo, List<WriteInfo>> partitionWritesByShardStrategyTimestamp(List<WriteInfo> writes) {
         int shards = numShards.get();
-        return writes.stream()
-                .filter(writeInfo -> getStrategy(writeInfo) != SweeperStrategy.NON_SWEEPABLE)
-                .collect(Collectors.groupingBy(write -> getPartitionInfo(write, shards), Collectors.toList()));
-    }
 
-    private PartitionInfo getPartitionInfo(WriteInfo write, int shards) {
-        return PartitionInfo.of(write.toShard(shards), getStrategy(write), write.timestamp());
+        // Note that we do a single pass over writes to determine their sweep strategy, and only add sweepable write to
+        // sweepablePartitionedWrites map, so when it is empty we consider all writes for that timestamp non-sweepable.
+        Map<PartitionInfo, List<WriteInfo>> sweepablePartitionedWrites = new HashMap<>();
+        for (WriteInfo writeInfo : writes) {
+            SweeperStrategy strategy = getStrategy(writeInfo);
+            if (strategy != SweeperStrategy.NON_SWEEPABLE) {
+                PartitionInfo partition = PartitionInfo.of(writeInfo.toShard(shards), strategy, writeInfo.timestamp());
+                sweepablePartitionedWrites
+                        .computeIfAbsent(partition, _k -> new ArrayList<>())
+                        .add(writeInfo);
+            }
+        }
+
+        if (sweepablePartitionedWrites.isEmpty()) {
+            // All writes for this partition are non-sweepable
+            long startTs = writes.stream().findFirst().orElseThrow().timestamp();
+            return SweepQueueUtils.partitioningForNonSweepable(startTs);
+        }
+        return sweepablePartitionedWrites;
     }
 
     @VisibleForTesting
     SweeperStrategy getStrategy(WriteInfo writeInfo) {
-        return writeInfo
-                .writeRef()
-                .map(WriteReference::tableRef)
-                .map(this::getStrategyForTable)
-                .orElse(SweeperStrategy.NON_SWEEPABLE);
+        Optional<WriteReference> writeReference = writeInfo.writeRef();
+        if (writeReference.isPresent()) {
+            return getStrategyForTable(writeReference.get().tableRef());
+        }
+        return SweeperStrategy.NON_SWEEPABLE;
```

Comment:
just for my reference: does doing things in this way end up being more efficient/doing fewer allocations than the 'chained Optionals' pipeline?

Reply:
yes, the chained `Optional::map` will allocate a new `Optional` for each invocation that returns a non-empty value. In ~99%+ of cases the `Optional` allocations won't matter but on really hot paths we're better off avoiding the allocation if we can either via checking present/empty or just dealing with `@Nullable` at the expense of potentially less concise code.

- https://github.com/openjdk/jdk/blob/jdk-17-ga/src/java.base/share/classes/java/util/Optional.java#L260
- https://github.com/openjdk/jdk/blob/jdk-17-ga/src/java.base/share/classes/java/util/Optional.java#L126-L128